// app/utils/svizzle/style.js
import {
	isIterableNotEmpty,
	makePostfixed,
	makePrefixed
} from '@svizzle/utils';
import colorParse from 'color-parse';
import * as _ from 'lamb';

import {makeTrimmedSplitBy} from './utils';

const parseCssText = _.pipe([
	_.splitBy(';'),
	_.filterWith(isIterableNotEmpty),
	_.mapWith(makeTrimmedSplitBy(':'))
]);

export const getStylesheet = href => _.find(
	[...document.styleSheets], // convert collection to array
	_.hasKeyValue('href', href)
);

export const makeGetStyleRulesObj = selectorRegex => _.pipe([
	_.filterWith(_.pipe([
		_.getKey('selectorText'),
		makeTrimmedSplitBy(','),
		_.some(selectorRegex.test.bind(selectorRegex))
	])),
	_.mapWith(_.collect([
		_.getKey('selectorText'),
		_.pipe([
			_.getPath('style.cssText'),
			parseCssText,
			_.fromPairs
		])
	])),
	_.fromPairs
]);

export const getAllStylesBySelector = _.pipe([
	_.mapWith(_.collect([
		_.getKey('selectorText'),
		_.getKey('style')
	])),
	_.fromPairs
]);

export const setStyleRules = (targetRules, srcRules) => {
	const selectors = _.keys(srcRules);
	selectors.forEach(selector => {
		if (selector in targetRules) {
			const style = srcRules[selector];
			const properties = _.keys(style);
			properties.forEach(prop => {
				targetRules[selector].setProperty(prop, style[prop])
			});
		}
	})
}

const hexValue = number => number.toString(16).padStart(2, '0');

export const getRGBColor = themeVars => _.pipe([
	_.curry(_.getIn)(themeVars),
	colorParse,
	_.getKey('values'),
	_.mapWith(hexValue),
	_.joinWith(''),
	makePrefixed('#')
]);

export const getRGBAColor = themeVars => _.pipe([
	_.curry(_.getIn)(themeVars),
	colorParse,
	_.collect([
		_.getKey('values'),
		_.pipe([
			_.getKey('alpha'),
			_.multiplyBy(255)
		])
	]),
	_.flatten,
	_.mapWith(hexValue),
	_.joinWith(''),
	makePrefixed('#')
]);

const getPropDef = _.pipe([
	_.joinWith(': '),
	makePostfixed(';\n')
]);
const getRuleDefs = _.pipe([
	_.pairs,
	_.mapWith(getPropDef),
	_.joinWith('\t'),
	makePrefixed('{\n\t'),
	makePostfixed('}')
]);
const getClassDef = _.pipe([
	_.collect([
		_.getAt(0),
		_.pipe([
			_.getAt(1),
			getRuleDefs
		])
	]),
	_.joinWith(' ')
]);
export const getThemeClassDefs = _.pipe([
	_.pairs,
	_.mapWith(getClassDef),
	_.joinWith('\n'),
]);
