// app/stores/interaction.js
import * as _ from 'lamb';
import {writable, get} from 'svelte/store';

import {_dataset} from 'app/stores/dataset';

export const _hero = writable(null);

export const clearHero = () => {
	_hero.set(null);
}

const setHero = ({id, isPinned, isVisible}) => {
	let hero;
	const orgsById = get(_dataset).orgsById;
	const org = orgsById[id];
	if (org) {
		hero = {
			isVisible,
			isPinned,
			org
		};
	}
	_hero.set(hero);
};

export const setFocusedHero = id => {
	!get(_hero)?.isPinned && setHero({
		id,
		isPinned:false,
		isVisible: false
	});
};

export const setFocusedHeroFromMap = id => {
	!get(_hero)?.isPinned && setHero({
		id,
		isPinned:false,
		isVisible: true
	});
};

export const setSelectedHero = id => {
	const hero = get(_hero);
	const isNotPinned = hero?.id === id && hero.isPinned;
	setHero({
		id,
		isPinned: !isNotPinned,
		isVisible: true
	});
};

export const clearInteractionStores = () => {
	_hero.update(hero => hero?.isPinned ? hero : null);
}

export const defaultFooterText = 'Click on background to dismiss';

export const _isCursorOnMap = writable();

export const setIsCursorOnMap = () => _isCursorOnMap.set(true);
export const clearIsCursorOnMap = () => _isCursorOnMap.set(false);
